### ISSUE TITLE
Upload stalling w/o reaching daily quota limits

### ISSUE THREAD

**Issue by snairolf** [ylexus/jiotty-photos-uploader#98]
My uploads are stalling and don't make any progress over several hours.

The logs show:
- com.google.api.gax.rpc.ResourceExhaustedException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: Quota exceeded for quota metric 'Write requests' and limit 'Write requests per minute per user' of service 'photoslibrary.googleapis.com' for consumer 'project_number_XXX'

**Comment by snairolf** [comment]
Here are some other errors and the exception stack trace:
2021-05-13T15:05:05,192 DEBUG [upload-pool-0] n.y.g.c.BackingOffRemoteApiExceptionHandlerImpl Retryable exception performing operation 'create media items', backing off by waiting for 1860ms
com.google.api.gax.rpc.InternalException: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:67) ~[gax-1.50.1.jar:1.50.1]
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72) ~[gax-grpc-1.50.1.jar:1.50.1]
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60) ~[gax-grpc-1.50.1.jar:1.50.1]
	at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97) ~[gax-grpc-1.50.1.jar:1.50.1]
	at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:68) ~[api-common-1.8.1.jar:?]
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1074) ~[guava-30.1.1-jre.jar:?]
	at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:30) ~[guava-30.1.1-jre.jar:?]
	at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1213) ~[guava-30.1.1-jre.jar:?]
	at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:983) ~[guava-30.1.1-jre.jar:?]
	at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:771) ~[guava-30.1.1-jre.jar:?]
	at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:526) ~[grpc-stub-1.29.0.jar:1.29.0]
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:501) ~[grpc-stub-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:426) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl.access$500(ClientCallImpl.java:66) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:689) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$900(ClientCallImpl.java:577) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:751) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:740) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[grpc-core-1.29.0.jar:1.29.0]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[grpc-core-1.29.0.jar:1.29.0]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) ~[?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
	Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
		at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57) ~[gax-1.50.1.jar:1.50.1]
		at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112) ~[gax-1.50.1.jar:1.50.1]
		at com.google.photos.library.v1.internal.InternalPhotosLibraryClient.batchCreateMediaItems(InternalPhotosLibraryClient.java:353) ~[google-photos-library-client-1.5.0.jar:1.5.0]
		at com.google.photos.library.v1.PhotosLibraryClient.batchCreateMediaItems(PhotosLibraryClient.java:176) ~[google-photos-library-client-1.5.0.jar:1.5.0]
		at net.yudichev.jiotty.connector.google.photos.GooglePhotosClientImpl.lambda$null$5(GooglePhotosClientImpl.java:88) ~[jiotty-connector-google-photos-1.6.5.jar:?]
		at java.util.Optional.map(Unknown Source) ~[?:?]
		at net.yudichev.jiotty.connector.google.photos.GooglePhotosClientImpl.lambda$null$8(GooglePhotosClientImpl.java:88) ~[jiotty-connector-google-photos-1.6.5.jar:?]
		at net.yudichev.jiotty.common.inject.BaseLifecycleComponent.lambda$whenStartedAndNotLifecycling$5(BaseLifecycleComponent.java:60) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.jiotty.common.lang.Locks.inLock(Locks.java:22) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.jiotty.common.inject.BaseLifecycleComponent.whenNotLifecycling(BaseLifecycleComponent.java:54) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.jiotty.common.inject.BaseLifecycleComponent.whenStartedAndNotLifecycling(BaseLifecycleComponent.java:58) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.jiotty.connector.google.photos.GooglePhotosClientImpl.lambda$createMediaItems$9(GooglePhotosClientImpl.java:82) ~[jiotty-connector-google-photos-1.6.5.jar:?]
		at java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.BackpressuredExecutorServiceProvider.rejectedExecution(BackpressuredExecutorServiceProvider.java:29) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.ThreadPoolExecutor.reject(Unknown Source) ~[?:?]
		at java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.asyncSupplyStage(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.supplyAsync(Unknown Source) ~[?:?]
		at net.yudichev.jiotty.connector.google.photos.GooglePhotosClientImpl.createMediaItems(GooglePhotosClientImpl.java:82) ~[jiotty-connector-google-photos-1.6.5.jar:?]
		at net.yudichev.googlephotosupload.core.GooglePhotosUploaderImpl.lambda$createMediaItems$11(GooglePhotosUploaderImpl.java:183) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.withBackOffAndRetry(CloudOperationHelperImpl.java:26) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$2(CloudOperationHelperImpl.java:41) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.Optional.map(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$4(CloudOperationHelperImpl.java:38) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.jiotty.common.lang.Either.map(Either.java:41) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$5(CloudOperationHelperImpl.java:35) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.withBackOffAndRetry(CloudOperationHelperImpl.java:35) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$2(CloudOperationHelperImpl.java:41) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.Optional.map(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$4(CloudOperationHelperImpl.java:38) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.jiotty.common.lang.Either.map(Either.java:41) ~[jiotty-common-1.6.5.jar:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.lambda$withBackOffAndRetry$5(CloudOperationHelperImpl.java:35) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.CloudOperationHelperImpl.withBackOffAndRetry(CloudOperationHelperImpl.java:35) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.GooglePhotosUploaderImpl.createMediaItems(GooglePhotosUploaderImpl.java:181) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.GooglePhotosUploaderImpl.lambda$uploadDirectory$5(GooglePhotosUploaderImpl.java:110) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.AddToAlbumWhileCreatingStrategy.lambda$addToAlbum$0(AddToAlbumWhileCreatingStrategy.java:25) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
		at java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
		at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
		at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.AddToAlbumWhileCreatingStrategy.lambda$addToAlbum$2(AddToAlbumWhileCreatingStrategy.java:26) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.AddToAlbumWhileCreatingStrategy.addToAlbum(AddToAlbumWhileCreatingStrategy.java:24) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.SelectingAddToAlbumStrategy.addToAlbum(SelectingAddToAlbumStrategy.java:42) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.GooglePhotosUploaderImpl.lambda$uploadDirectory$6(GooglePhotosUploaderImpl.java:106) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.CompletableFuture.uniComposeStage(Unknown Source) ~[?:?]
		at java.util.concurrent.CompletableFuture.thenCompose(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.GooglePhotosUploaderImpl.uploadDirectory(GooglePhotosUploaderImpl.java:91) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at net.yudichev.googlephotosupload.core.UploaderImpl.lambda$upload$1(UploaderImpl.java:64) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source) ~[?:?]
		at java.util.Spliterators$ArraySpliterator.forEachRemaining(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.copyInto(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source) ~[?:?]
		at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(Unknown Source) ~[?:?]
		at java.util.stream.AbstractPipeline.evaluate(Unknown Source) ~[?:?]
		at java.util.stream.ReferencePipeline.collect(Unknown Source) ~[?:?]
		at net.yudichev.googlephotosupload.core.UploaderImpl.lambda$upload$4(UploaderImpl.java:70) ~[Jiotty%20Photos%20Uploader-1.8.0.jar:1.8.0]
		at java.util.concurrent.CompletableFuture$UniCompose.tryFire(Unknown Source) [?:?]
		at java.util.concurrent.CompletableFuture.postComplete(Unknown Source) [?:?]
		at java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source) [?:?]
		at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
		at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: io.grpc.StatusRuntimeException: INTERNAL: Internal error encountered.
	at io.grpc.Status.asRuntimeException(Status.java:533) ~[grpc-api-1.29.0.jar:1.29.0]
	... 15 more
2021-05-13T15:05:07,056 DEBUG [upload-pool-0] n.y.g.c.CloudOperationHelperImpl Retrying operation 'create media items' with backoff Optional[1860]ms

**Comment by ylexus** [comment]
I’ll need full logs for the session from start in order to analyse it effectively. Could you email it to me please (you know my address by now)? Ideally also config.json from the same directory where the logs for is located. 

But overall if you see RESOURCE_EXHAUSTED exceptions happening even overnight, then there’s nothing we can do. It should still make some progress, but slow. They have various limits as you can see, like requests per minute, requests per day etc.

**Comment by snairolf** [comment]
I'll look into sharing the logs. There is no config.json in that directory, just a data.json.

I am wondering if the back-off is long enough to prevent hitting the per minute quota limit. It really doesn't seem like there much progress, maybe 3 images over several hours.

**Comment by ylexus** [comment]
Sorry, data.json, yes. The back off is exponential. But I need to see the progress of the back off and for that I need logs.

**Comment by ylexus** [comment]
One thing that I can see is that RESOURCE_EXHAUSTED (429) error should be retried with a min of 30 second delay as per https://developers.google.com/photos/library/guides/best-practices#retrying-failed-requests, but I am not distinguishing between kinds of retryable errors. Something that I'll improve.

**Comment by ylexus** [comment]
Done - if you're building from sources, pull and try the new version.

**Comment by ylexus** [comment]
All we know about the problem so far is that it's a specific picture that gets stuck with INTERNAL GRPC error:
```
2021-05-15T22:02:29,951 DEBUG [upload-pool-0] n.y.j.c.g.p.GooglePhotosClientImpl Create media items in album Optional[REMOVED], items [NewMediaItem{uploadToken=REMOVED, fileName=DSC_3328-original.JPG, description=Optional[DSC_3328-original.JPG]}]
2021-05-15T22:02:29,953 DEBUG [grpc-nio-worker-ELG-1-2] i.g.n.s.i.n.u.i.l.AbstractInternalLogger [id: 0xad824423, L:/192.168.0.150:54067 - R:photoslibrary.googleapis.com/142.250.72.138:443] OUTBOUND HEADERS: streamId=5 headers=GrpcHttp2OutboundHeaders[:authority: photoslibrary.googleapis.com:443, :path: /google.photos.library.v1.PhotosLibrary/BatchCreateMediaItems, :method: POST, :scheme: https, content-type: application/grpc, te: trailers, user-agent: grpc-java-netty/1.29.0, x-goog-api-client: gl-java/14.0.2 gapic/1.5.0 gax/1.50.1 grpc/1.29.0, grpc-accept-encoding: gzip, authorization: Bearer REMOVED, grpc-timeout: 19999776u] streamDependency=0 weight=16 exclusive=false padding=0 endStream=false
2021-05-15T22:02:29,954 DEBUG [grpc-nio-worker-ELG-1-2] i.g.n.s.i.n.u.i.l.AbstractInternalLogger [id: 0xad824423, L:/192.168.0.150:54067 - R:photoslibrary.googleapis.com/142.250.72.138:443] OUTBOUND DATA: streamId=5 padding=0 endStream=true length=648 bytes=00000002830a4c41484d44534c5361614d547a376652756f543348596141665776494c7631566737785a7a39434337665a2d6d6a6e626b6b5f4333534f484179...
2021-05-15T22:02:31,001 DEBUG [grpc-nio-worker-ELG-1-2] i.g.n.s.i.n.u.i.l.AbstractInternalLogger [id: 0xad824423, L:/192.168.0.150:54067 - R:photoslibrary.googleapis.com/142.250.72.138:443] INBOUND HEADERS: streamId=5 headers=GrpcHttp2ResponseHeaders[:status: 200, content-type: application/grpc, grpc-accept-encoding: identity,deflate,gzip, accept-encoding: identity,gzip, content-length: 0, date: Sat, 15 May 2021 22:02:31 GMT, alt-svc: h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"] padding=0 endStream=false
2021-05-15T22:02:31,001 DEBUG [grpc-nio-worker-ELG-1-2] i.g.n.s.i.n.u.i.l.AbstractInternalLogger [id: 0xad824423, L:/192.168.0.150:54067 - R:photoslibrary.googleapis.com/142.250.72.138:443] INBOUND HEADERS: streamId=5 headers=GrpcHttp2ResponseHeaders[grpc-status: 13, grpc-message: Internal error encountered., grpc-server-stats-bin: AABKPDc6AAAAAA] padding=0 endStream=true
```

grpc-status: 13, grpc-message: Internal error encountered

**Comment by ylexus** [comment]
Note to self: check that the "Some of the files I will upload..." option still works for:
1. Files uploaded by other app
    1. Not present in any album
    2. Present in an album created by Jiotty
    3. Present in an album created by another app
2. Files uploaded by Jiotty but later added to an album created by other app

It may be that the loophole this option was using no longer works.

**Comment by ylexus** [comment]
All scenarios that I can think of with a pre-existing album and/or image pre-uploaded manually - they all still work, so this problem was not due to any API permission changes on Google side that I suspected. Oh well, a least the minor enhancements that I made would make it hang forever if this ever happens again. There is nothing else I can think of. I wish Google server responded with a better error than "INTERNAL" to give us some more clues. I'll close for now.
